### PR TITLE
BUG: stats: preserve sample dtype for bin edges

### LINKED ISSUE #13587: binned_statistic unreliable with single precision
When using single precision values not every value in my array is put into a bin by `binned_statistic`. The final (closed) bin interval excludes my original single precision value.

#### Reproducing code example:

```
from scipy.stats import binned_statistic
import numpy

x = numpy.array([0, 0.42358226], dtype=numpy.float32)
count, _, _ = binned_statistic(x, values=None, statistic='count')
count.sum()!=x.shape[0]
```

#### Error message:
The second data point in my data array is not included in any bin.

#### Scipy/Numpy/Python version information:
```
1.6.1 1.19.4 sys.version_info(major=3, minor=7, micro=8, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by jeffplourde**
#### Reference issue
Closes gh-13587.

#### What does this implement/fix?
I ran across a case in which the max value of a single precision sample gets excluded from the right-most bin in `binned_statistic`. I propose maintaining the same `dtype` for the `edges` as in the `sample` by specifying the `dtype` kwarg to the `linspace` or `asarray` calls that create the `edges` array.

#### Additional information
The sample value used in my unit test is the one I discovered empirically.

### CODE REVIEW

**Comment by mdhaber** [comment]
Thanks @jeffplourde. Looks simple and effective.

Take a look at
image

by clicking Details and following the red Xs. The data type of `bin_edges` changed, so they're represented as integers now, apparently.
```
2021-02-20T02:40:57.3418896Z ===========
2021-02-20T02:40:57.3419280Z scipy.stats
2021-02-20T02:40:57.3419644Z ===========
2021-02-20T02:40:57.3419893Z 
2021-02-20T02:40:57.3420274Z scipy.stats.binned_statistic
2021-02-20T02:40:57.3421337Z ----------------------------
2021-02-20T02:40:57.3421632Z 
2021-02-20T02:40:57.3422418Z File "build/testenv/lib/python3.8/site-packages/scipy/stats/_binned_statistic.py", line 107, in binned_statistic
2021-02-20T02:40:57.3423033Z Failed example:
2021-02-20T02:40:57.3423738Z     stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)
2021-02-20T02:40:57.3424228Z Expected:
2021-02-20T02:40:57.3424734Z     BinnedStatisticResult(statistic=array([4. , 4.5]),
2021-02-20T02:40:57.3425547Z             bin_edges=array([1., 4., 7.]), binnumber=array([1, 1, 1, 2, 2]))
2021-02-20T02:40:57.3426040Z Got:
2021-02-20T02:40:57.3426574Z     BinnedStatisticResult(statistic=array([4. , 4.5]), bin_edges=array([1, 4, 7]), binnumber=array([1, 1, 1, 2, 2]))
2021-02-20T02:40:57.3426993Z 
2021-02-20T02:40:57.3427203Z 
2021-02-20T02:40:57.3427995Z File "build/testenv/lib/python3.8/site-packages/scipy/stats/_binned_statistic.py", line 115, in binned_statistic
2021-02-20T02:40:57.3428698Z Failed example:
2021-02-20T02:40:57.3429407Z     stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)
2021-02-20T02:40:57.3429889Z Expected:
2021-02-20T02:40:57.3430344Z     BinnedStatisticResult(statistic=array([[4. , 4.5],
2021-02-20T02:40:57.3430879Z            [8. , 9. ]]), bin_edges=array([1., 4., 7.]),
2021-02-20T02:40:57.3431376Z            binnumber=array([1, 1, 1, 2, 2]))
2021-02-20T02:40:57.3431803Z Got:
2021-02-20T02:40:57.3432224Z     BinnedStatisticResult(statistic=array([[4. , 4.5],
2021-02-20T02:40:57.3432803Z            [8. , 9. ]]), bin_edges=array([1, 4, 7]), binnumber=array([1, 1, 1, 2, 2]))
2021-02-20T02:40:57.3433161Z 
2021-02-20T02:40:57.3433369Z 
2021-02-20T02:40:57.3434253Z File "build/testenv/lib/python3.8/site-packages/scipy/stats/_binned_statistic.py", line 120, in binned_statistic
2021-02-20T02:40:57.3434810Z Failed example:
2021-02-20T02:40:57.3435462Z     stats.binned_statistic([1, 2, 1, 2, 4], np.arange(5), statistic='mean',
2021-02-20T02:40:57.3435976Z                            bins=3)
2021-02-20T02:40:57.3436536Z Expected:
2021-02-20T02:40:57.3436993Z     BinnedStatisticResult(statistic=array([1., 2., 4.]),
2021-02-20T02:40:57.3437488Z             bin_edges=array([1., 2., 3., 4.]),
2021-02-20T02:40:57.3438004Z             binnumber=array([1, 2, 1, 2, 3]))
2021-02-20T02:40:57.3438430Z Got:
2021-02-20T02:40:57.3438957Z     BinnedStatisticResult(statistic=array([1., 2., 4.]), bin_edges=array([1, 2, 3, 4]), binnumber=array([1, 2, 1, 2, 3]))
```

The refguide checks are very picky, so this isn't inherently a problem. It might be OK to just remove the decimal places from the examples. But we need to think through whether there are other implications that could cause other problems. Any guesses why the original author might have specified that they be `float`s?

**Comment by jeffplourde** [comment]
Thanks for pointing out the refguide checks. I wasn't aware of them.

Maybe the original author was going for equivalency with `np.linspace` absent a `dtype` kwarg; from the numpy docs:
> The inferred dtype will never be an integer; float is chosen even if the arguments would produce an array of integers.

I hesitate to change the behavior so I propose only maintaining the dtype if the sample is floating point. In this way only callers passing floating point values that are not double precision should see a change in behavior; and for the better as their bin edges will not have more (or less) precision than their sample data. Maybe it's a lack of imagination but I can't foresee a case where having a type mismatch between your floating point sample and your floating point bin edges would be desirable.

**Review by mdhaber — APPROVED** [review]
I think it was a good idea to make this change only for subtypes of `np.floating`. I am not able to predict any unintended consequences, and `binned_statistic` seems pretty well tested, so it looks good to me, but I think this should get another pair of eyes before merging.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @jeffplourde and @mdhaber.

It looks fine to me to only preserve floating-point precision, and avoid returning integers, for backwards compatibility.

**Pull request merged**
